[Draft] XGS1210 support and more RTL9300 features (v3)

oliv3r opened the pull request:
This PR is a continuation of #10319.



This PR improves the feature support for the RTL9300 platform, specifically it adds:

 *  SFP+ support for fibre and ethernet modules (10G and 1G)
 *  Improved RTL8226 and RTL8218d support for the XGS models
 *  LED setup for the Zyxel XGS models without the need for u-boot
 *  VLAN fixes
 *  EEE improvements on all platforms and all supported PHYs
 *  EEE support for the 1G and 2.5G ports of the XGS 1210-12 and for the 1G ports on the XGS1250
 *  LDPS support for all platforms

These features are then used to add support for the Zyxel XGS1210-12 (and XGS1010-12), which has complete HW support (1GBit, 2.5GBit ports, SFP+, EEE, LDPS)

For the best user-experience, it is recommended to also use a re-compiled, slightly modified U-Boot from the realtek-SDK. The source and binary can be found on [gitlab.com](https://gitlab.com/olliver/openwrt/realtek_sdk) with a forum thread on the [openwrt forums](https://forum.openwrt.org/t/u-boot-for-xgs12xx-switches-and-others-based-on-realtek-rtl930x/134478).

Commit:
realtek: Fix CONFIG_MDIO_SMBUS dependency on RTL83xx

CONFIG_SFP depends on CONFIG_MDIO_SMBUS now and this depends on
CONFIG_I2C_SMBUS.
This created the following warnings on the RTL83xx platforms:
WARNING: unmet direct dependencies detected for MDIO_SMBUS
  Depends on [n]: NETDEVICES [=y] && MDIO_DEVICE [=y] && MDIO_BUS [=y] && I2C_SMBUS [=n]
  Selected by [y]:
  - SFP [=y] && NETDEVICES [=y] && PHYLIB [=y] && I2C [=y] && PHYLINK [=y] && (HWMON [=y] || HWMON [=y]=n)

We remove the overall dependency to CONFIG_MDIO_SMBUS from CONFIG_SFP and
instead add it to CONFIG_I2C_RTL9300 since only the RTL93xx platforms make
use of the SMBus capable I2C controller.

Reported-by: Hauke Mehrtens <hauke@hauke-m.de>
Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Introduce SoC specific Switch IRQ settings

Add SoC specific switch IRQ settings for media change and monitoring
of the SerDes.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek/rtl930x: add support for Zyxel XGS1210-12

The Zyxel XGS1210-12 is an 8x 1 GBit, 2x 2.5 GBit NBaseT ports and
2 SFP+ ports switch.

Hardware:
	- RTL9302 SoC: MIPS 34Kc @ 800MHz
	- 128 MByte DRAM
	- 16MBit NOR Flash MX25L12833F
	- RTL8218D 8x GBit PHY
	- 2x RTL8226 2.5 GBit PHYs
	- 2x SFP+ Slots
	- 1 Reset button
	- Externally accessible UART header at the right side:
		Top-down: Vcc (3.3V), TX, RX, GND 115200 8N1
	- System LED

Installation:
	Connect to the Serial Port, power on the device
	Interrupt boot by pressing <ESC>
	Enable network by entering
	 > rtk network on
	Upload initramfs via tftp into RAM:
	 > tftpboot 0x84f00000 192.168.1.100:openwrt-realtek-rtl930x-zyxel_xgs1210-12-initramfs-kernel.bin
	Start initramfs
	 > bootm
	Once OpenWRT is running, install the squashfs image to flash.

Two images are available, an 'ideal' and preferred one, which does
require a custom U-Boot binary (available from source), and one that
keeps with the vendor partition layout, which does not offer use, but
does cost space. The later is preferred for users who do not wish to
re-flash their bootloader and wish to keep backwards compatibility at
the cost of about ~2 MiB worth of flash.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Only include used ports in IRQ masks

Use only used ports in interrupt masks for SDS/link faults
as otherwise these IRQs constantly fire.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: rtl93xx: Remove dead code

Remove unused code from rtl83xx_phylink_mac_link_state
which handled a case for the RTL9300 speed settings.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Adapt MAC-PHY connection for HiSGMII ports

On HiSGMII ports of the RTL9300 (used for the RTL8226), the
link between MAC and PHY is not statically HISGMII. Instead
for speeds lower than 2.5G, SGMII is used. Detect this
change, reconfigure the SerDes to use that mode and then
recalibrate that link.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Add support for SFP+ ports on RTL9300

This adds support for SFP and SFP+ modules in RTL9300
based switches. For this we add configuration code
for the RTL9300 SerDes in order to support 1000Base-X
and 10G PHY modes.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Use algorithm to calculate LED counts, instead of heuristics

This counts the number of LEDs in a LED-set and sets the LED-count
for a given port based on its configured LED-set instead of the
type of port.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Limit link failure interrupts

To avoid flooding the system with link-failure interrupts, lets limit
them.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Only read link-status once

The link status register returns information for all ports, no need to
read it out for each individual port.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

splitme

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Fix DSA handling of SFP+ ports on RTL9300

This fixes handling of the SFP+ ports in DSA by correcting
the output of the link-state routine.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Disable RX symbol error and PHY-update IRQs

As we are not able to do something sensible with them.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix handling of USXGMII links on RTL9300

USXGMII links are 10G links and not also 2.5GBit links.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix SFP+ port in Zyxel XGS1250-12

The SFP+ port is a managed port which does not have a fixed 10Gbit link
but can also hold 1000Base-X modules. Fix this in the .dts.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Improve SMI polling configuration based on .dts

Improves error detection of the association of smi addresses to ports
on RTL93xx. Allows empty smi addresses for ports using a PHY linked to an
internal SerDes.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Fix stripping of VLAN tags on RTL93xx

On the RTL93xx platforms, the VLAN tags were stripped from egressing
and ingressing packets on a port. Keep all these tags intact.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Fix MAC port up/down configuration on RTL9300

The MAC link configuration on the RTL9300 was based on a misunderstanding
of how the registers worked. Fix this.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Add port mirroring support for RTL93xx

This adds support for port mirroring on the RTL93xx platforms.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix reset register access

The reset register on RTL93xx not merely have bits to execute
a reset of a hardware component, but also configuration bits for
reset procedures. Keep them during executing a reset.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Fix bugs in MMD-register access on RTL838x

This fixes 2 bugs in MMD access on the RTL838x:
	- We need to preserve the state of the park-page because the
	  normal PHY r/w functions expect a correct setting
	  in RTL838X_SMI_ACCESS_PHY_CTRL_1
	- commit b53202a8c3f728c348c5376e5b5fb36af7c37744 introduced
	  calling the hw MMD access functions with mdiobus_c45_devad(regnum)
	  so now we need to invert this op via mdiobus_c45_regad(reg)

This was tested with MMD access register R/W for EEE functionality
on the Zyxel GS1900-10HP.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: improve EEE handling

Instead of setting eee->advertised at the MAC-layer we leave this now
to the PHYs to set correctly. We also introduce reading out the LPI
(Low Power Idle) timer.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix HW route lookup

Fix various bugs found in the HW route lookup function.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Add support for Link Down Power Saving

Adds support for RTL8218b, RTL8218d and RTL8226 for LDPS
which reduces power consumption while a link is down.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Add support for RTL8214QF

This adds support for the Realtek RTL8214QF PHY, which has a single 5Gbit
QSGMII link and provides 4 1000BX fiber connections for use with SFP
ports. The RTL8214QF PHY is also the HW base for the 10GBit SFP+
PHYs found in the RTL8396 and RTL93xx SoCs.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Improve RTL9300 SerDes initialization

This improves RTL9300 SerDes initialization, including configuration
settings for different link modes to the MAC.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Improve debugging information to aid development

Print beter debugging information to help with diagnosing problems with
the newly introduced RTL8226 work.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Add configuration support for the RTL8226

This adds support for the configuration of the RTL8226 PHYs, including
support for HiSGMII and SGMII connections used with the RTL9300 SerDes
for 2.5GBit and 1GBit modes. Setup includes TX configuration and RX
calibration of the PHY-MAC link. We provide for MDIO pin swapping as found
in the Zyxel XGS1210 switch.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Use regular PHY registers defines

These PHY register from the proper header should be used instead.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Improve RTL8218D configuration

This improves the RTL8218D configuration to make use of the
new setup of the Realtek RTL9300 SerDes including calibration support.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Commit:
realtek: Improve EEE suport in PHY drivers

Improve the EEE support in the PHY drivers of the RTL8218B/D/8826:
 - calculate eee->advertised from MMD values
 - enable/disable EEE via MDIO_MMD_AN.MDIO_AN_EEE_ADV where possible
 - support for 2.5 GBit EEE advertisment/LP report on RTL8226

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

oliv3r committed:
realtek/rtl930x: Add shared base for xgs12xx series

The XGS1250-21, XGS1210-12 and XGS1010-12 are very similar, with very
minor differences. Lets have a shared devicetree to make maintaining
multiple devices with a common base easier.

The new shared devicetree also drops the compatible partition layout
with the OEM firmware, as it does not serve any purpose and only wastes
space.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek/rtl930x: Migrate XGS1250-12 to common devicetree

With the newly introduced shared devicetree for the XGS12xx series, also
switch the XGS1250-12 to use the shared devicetree.

While here, clean up a small ugliness, without changing actual
functionality.

Note, that due to the weird I2C muxing, there's some overlap there that
will be addressed in the future with a proper I2C driver that properly
handles this.

This change also includes a build to keep vendor U-Boot/flash layout
compatibility. The preferred method is to use the update U-Boot SPL.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek/rtl930x: add support for Zyxel XGS1010-12

The Zyxel XGS1010-12 is an 8x 1 GBit, 2x 2.5 GBit NBaseT ports and
2 SFP+ ports switch. It is identical to the Zyxel XGS1210-12, except for
the partition layout which is probably caused by the basic firmware
coming from the SDK with the default dual partition layout. To ensure
backwards compatibility and more importantly to those who also do not
wish to replace their bootloader, a second image type with the postfix
`_vendor` is also added. Using our custom U-Boot adds about ~7MiB of
storage to the firmware image compared to stock.

In addition to the above, the XGS1010-12 lacks a physical reset button,
but all circuitry is in place to support this, including the whole in
the metal housing. If a switch is added at a later time, the switch is
basically converted to an XGS1210-12 and thus that firmware should be
used instead.

As this switch does not feature any form of management or firmware
updateability, one must use the serial UART either through XYZ-modem
(untested) or tftp. The instructions to do so are identical as with the
XGS1250-12.

When using our customized U-Boot, `upgrade runtime` can be used
instead over ftp as well.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
cosmetic: realtek/rtl9302: Whitespace cleanup XGS1250-12

No code changes are part of this comment.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek/rtl930x: Use LZMA compression for XGS1250-12

The XGS1250-12 bootloader has LZMA compression support and stock
firmware is LZMA compressed. Lets do the same to save some space.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

Neustradamus commented:
To follow this PR :)

oliv3r commented:
I've dropped the EEE stuff from my local branch and will test that, and check if I can drop some other 'not crucial' patches to make this MR smaller.

But birger went through a lot of trouble to add support for these new series of switches; and 'getting something' in the hands of users, is probably still very valueable :) More exposure would be good.

janh commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/common.c:
```diff
@@ -320,22 +320,20 @@ static int __init rtl83xx_mdio_probe(struct rtl838x_switch_priv *priv)
 				priv->ports[pn].sds_num);
 		}
-		pr_info("%s port %d has SDS\n", __func__, priv->ports[pn].sds_num);
+		pr_debug("%s port %d has SDS %d\n", __func__, pn, priv->ports[pn].sds_num);
```
This change also happens to solve the watchdog-related issue on some HPE 1920 devices. (I didn't test with debug messages enabled during boot, though, so adding something like `cond_resched` additionally may still be worth considering.)

Maybe it would make sense to split out all the non-device specific commits into separate pull requests?

oliv3r replied:
@janh please do a (github) suggestion with the `cond_resched` and I'll put it in.

As for your fixes and non-device specific fixes, Maybe/probably. I'm still cleaning up the patches, re-organizing them. I don't want to mess with the patches content wise too much. But there's 3 goals for me personally. My primary goal, get my switch support merged :) But for that, secondly, I need the 9300 enablement patches. Not sure if we can merge patches without any devices.

Thirdly, the churn for getting stuff merged in openwrt is quite slow; so having 3 MR's (bugfixes/cleanups, 9300 support, device support) might mean foreverness, even though it's the right approach of course). Also, I lack any non-9300 hardware, so validating and testing the 'fixes' would be very hard.

lucize replied:
I do have an Zyxel XGS1250-12-ZZ0101F, maybe I'll open it and see about this, guess I can run initramfs

oliv3r replied:
initramfs would already be very helpful!

oliv3r replied:
> so adding something like cond_resched additionally may still be worth considering.

I've been thinking more about this, so the pr_debug is probably just the trigger here, it causes more cpu load then expected. (duh) but then what really IS the culprit? a race/timeout condition in this piece of code? I'm happy the change of the pr_info fixes it; i'm not happy that I don't yet understand why.

svanheule replied:
Are pr_debug() lines printed? If console output is disabled during boot, that can lead to considerable time (CPU cycle) savings.

oliv3r replied:
afaik not by default; you need to enable debug prints specifically (might even be during compile time).

I'd be more then happy to remove 80% of all those prints; but, as you know this is not my code :( so having some debugging will probably still be usefull for the forseable time; and boot-time is probably not that important atm?

janh replied:
> I've been thinking more about this, so the pr_debug is probably just the trigger here, it causes more cpu load then expected. (duh) but then what really IS the culprit? a race/timeout condition in this piece of code? I'm happy the change of the pr_info fixes it; i'm not happy that I don't yet understand why.

This is just because `printk` can be really slow, as it [directly writes to the console, if possible](https://www.kernel.org/doc/html/latest/core-api/printk-basics.html#c.printk).

On the HPE devices, the console runs at 38400 baud (this is also the rate used by the bootloader). Before this change, there are 3 calls to `pr_info` for every port, and if you calculate how long it takes just to write all of that to the console the result matches the actual execution time of this loop pretty well.

> afaik not by default; you need to enable debug prints specifically (might even be during compile time).

Yes, by default `pr_debug` is a no-op. The kernel needs to be built with `DEBUG` defined or with the [dynamic debug feature](https://www.kernel.org/doc/html/latest/admin-guide/dynamic-debug-howto.html) enabled.

I just did a few tests on my HPE 1920-48G with `CONFIG_KERNEL_DYNAMIC_DEBUG`, and `dyndbg="file drivers/net/dsa/rtl83xx/common.c +p"` added to cmdline to enable all debug prints from that file. I also tested with the second VPE disabled (`CONFIG_MIPS_MT_SMP` / `CONFIG_NR_CPUS` in the kernel config), as that gives more predictable results.

In these tests, I haven't seen the watchdog being triggered by this loop. The debug prints take about 0.7 seconds, so there is still a chance that it could happen if the watchdog hasn't been serviced shortly before.

The debug output from `rtl83xx_netdevice_event` together with that from `rtl83xx_fib_event` actually was enough to occasionally trigger the watchdog, though. And I wouldn't be surprised if enabling additional debug prints would show even more cases of this issue.

(I think if the [patch series moving the actual console writes to separate kernel threads](https://lore.kernel.org/linux-kernel/YouKQw72H7y9EJQK@alley/) actually lands in the kernel at some point, that would fix any issues related to slow `printk`. It was originally merged for Linux 5.19, but ended up being reverted as it didn't work properly with some console drivers.)

oliv3r replied:
You'd think that a) there'd be a debug print, that would do cond_reschedule after printing a few characters ;)

but even so, that loop must be 'quite full' without the debug prints? but then the question is; where to cond_reschedule? just before the pr_debug? Then it might be worth to add a local macro sched_pr_debug; that does cond_reschedule(); pr_debug instead.

Tricky business for sure regardless. Can we not set the console speed to 115200 on those devices as well? Yes it is annoying that U-Boot and linux have different speeds; but better then having a slow console?

Ideally, just replace all those custom u-boots with mainline u-boot and proper console speeds :) a man can dream ...

janh replied:
> but even so, that loop must be 'quite full' without the debug prints

I'm not sure, the `for_each_node_by_name(dn, "port")` loop just reads some data from the device tree. It seems to take about 0.45 ms to run for 52 ports here.

> but then the question is; where to cond_reschedule? just before the pr_debug? Then it might be worth to add a local macro sched_pr_debug; that does cond_reschedule(); pr_debug instead.

In the end this isn't really an issue specific to the switch driver, but a general kernel issue. I tested with all debug output in the kernel enabled (using `dyndbg="file * +p"`), and that triggered the watchdog just 3 seconds after it was initialized…

> Can we not set the console speed to 115200 on those devices as well? Yes it is annoying that U-Boot and linux have different speeds; but better then having a slow console?

I don't think this is a good option, as it really is quite annoying in practice. And there wouldn't be a clear separation between bootloader and Linux, as early Linux console output uses the same rate as the bootloader anyway.

oliv3r replied:
in short; we should add that cond_reschedule, in combination with that pr_debug (and a comment as to why then).
```suggestion
		/* The pr_debug below adds a lot of overhead, so as long as this pr is here, we need a reschedule */
		cond_reschedule();
		pr_debug("%s port %d has SDS %d\n", __func__, pn, priv->ports[pn].sds_num);
```

oliv3r committed:
cosmetic: realtek: Whitespace cleanup

Fix some ugly whitepsaces around the realtek sources.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

lucize commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -2952,13 +3107,594 @@ int rtl9300_sds_cmu_band_get(int sds)
+
+static u32 rtl8218d_serdes_model_get(struct phy_device *phydev)
+{
+	u32 block
```
semicolon missing

oliv3r replied:
yeah, sorry I pushed an intermediate state into this MR that I shouldn't have. Fixed now though.

lucize commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -2952,13 +3107,594 @@ int rtl9300_sds_cmu_band_get(int sds)
+	{ 0x28, 0x13, 0x027f }, { 0x28, 0x14, 0x1311 }, { 0x28, 0x16, 0x00c9 },
+	{ 0x28, 0x17, 0xa100 }, { 0x28, 0x1a, 0x0001 }, { 0x28, 0x1c, 0x0400 },
+	{ 0x29, 0x00, 0x820f }, { 0x29, 0x01, 0x0300 }, { 0x29, 0x02, 0x1017 },
+	{ 0x29, 0x03, 0xffdf }, { 0x29, 0x05, 0x7f7c }, { 0x29, 0x07, 0x8100 }
```
comma missing

lucize commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -2927,6 +3075,13 @@ sds_config rtl9300_a_sds_10gr_lane1[] =
+/* TODO: Add patch for USXGMII on SDS 6-8 for AQR113C */
+void rtl9300_sds_patch(int sds_num, sds_config c[], int entries)
+{
+	for (int i = 0; i < entries; ++i)
```
int i;
then remove the int from for loop

oliv3r replied:
actually, modern kernel development prefers he int loops :) (but only since 5.17)

lucize commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -2952,13 +3107,594 @@ int rtl9300_sds_cmu_band_get(int sds)
+
+int rtl9300_configure_serdes(int port, int sds_num, phy_interface_t phy_mode)
 {
 	struct device *dev = &phydev->mdio.dev;
```
I get `error: 'phydev' undeclared (first use in this function)`

lucize commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -2952,13 +3107,594 @@ int rtl9300_sds_cmu_band_get(int sds)
 	struct device_node *dn;
 	u32 sds_num = 0;
-	int sds_mode, calib_tries = 0, phy_mode = PHY_INTERFACE_MODE_10GBASER, i;
+	int sds_mode, calib_tries = 0, phy_mode = PHY_INTERFACE_MODE_10GBASER;
```
I get `error: 'phy_mode' redeclared as different kind of symbol`

`note: previous definition of 'phy_mode' with type 'phy_interface_t'`

lucize reviewed (commented):
so for 930x target I can't build

lucize commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -3784,7 +4516,7 @@ static int rtl8218b_int_phy_probe(struct phy_device *phydev)
 		shared->name = "RTL8218B (internal)";
 		/* Configuration must be done while patching still possible */
-		return rtl8380_configure_int_rtl8218b(phydev);
+		ret = rtl8380_configure_int_rtl8218b(phydev);
```
I get `error: 'ret' undeclared (first use in this function); did you mean 'net'?`

lucize commented:
`drivers/net/phy/rtl83xx-phy.c: In function 'rtl9300_sds_patch':
drivers/net/phy/rtl83xx-phy.c:3394:9: error: 'for' loop initial declarations are only allowed in C99 or C11 mode
 3394 |         for (int i = 0; i < entries; ++i)
      |         ^~~
drivers/net/phy/rtl83xx-phy.c:3394:9: note: use option '-std=c99', '-std=gnu99', '-std=c11' or '-std=gnu11' to compile your code
`
as I said, 

```
int i;
for (i = 0; i < entries; ++i)
```

the rest seems ok, I'll to test it later

oliv3r commented:
For now, I left all 'optional' patches at the tail of this series, but I'm more then happy to drop those and offer them in a seperate MR if that helps speeding up things.

oliv3r committed:
realtek: Expose rtl9300 sds functions

To be able to use the SDS functions elsewhere, expose their prototypes.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented:
@lucize current state seems to be working on 1 and 2.5G ports without the repetition

lucize commented:
still the same issue

working

```
[23546.576021] rtl83xx-switch switch@1b000000 lan9: Link is Down
[23546.583475] switch: port 9(lan9) entered disabled state
[23550.966547] rtl930x_switch_irq link faults: 09ffff00
[23550.972078] rtl930x_switch_irq RX symbol errors: 000001c4
[23551.769784] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 2500, link 1
[23551.779307] rtl93xx_phylink_mac_config SDS is 6
[23551.784450] rtl930x_switch_irq link faults: 09ffff00
[23551.789964] rtl930x_switch_irq RX symbol errors: 000001c4
[23551.796037] rtl83xx-switch switch@1b000000 lan9: Link is Up - 2.5Gbps/Full - flow control rx/tx
[23551.807027] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 2500, link 1
[23551.816646] rtl93xx_phylink_mac_config SDS is 6
[23551.822628] switch: port 9(lan9) entered blocking state
[23551.828533] switch: port 9(lan9) entered forwarding state
```

not working

```
[   66.640000] rtl930x_switch_irq link faults: 0dffff00
[   66.640000] rtl930x_switch_irq RX symbol errors: 000001c4
[   66.650000] rtl83xx-switch switch@1b000000 lan9: Link is Down
[   66.660000] switch: port 9(lan9) entered disabled state
[   67.610000] rtl9300_read_status: Port 27, SerDes is 9
[   67.610000] rtl9300_read_status got SDS mode 1f
[   68.640000] rtl9300_read_status: Port 27, SerDes is 9
[   68.640000] rtl9300_read_status got SDS mode 1f
[   69.500000] rtl930x_switch_irq link faults: 09ffff00
[   69.500000] rtl930x_switch_irq RX symbol errors: 000001c4
[   69.690000] rtl9300_read_status: Port 27, SerDes is 9
[   69.700000] rtl9300_read_status got SDS mode 1f
[   69.760000] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 1000, link 1
[   69.770000] rtl93xx_phylink_mac_config SDS is 6
[   69.780000] rtl930x_switch_irq link faults: 09ffff00
[   69.780000] rtl930x_switch_irq RX symbol errors: 000001c4
[   69.790000] rtl83xx-switch switch@1b000000 lan9: Link is Up - 1Gbps/Full - flow control rx/tx
[   69.800000] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 1000, link 1
[   69.810000] rtl93xx_phylink_mac_config SDS is 6
[   69.810000] switch: port 9(lan9) entered blocking state
[   69.820000] switch: port 9(lan9) entered forwarding state
[   70.720000] rtl9300_read_status: Port 27, SerDes is 9
[   70.720000] rtl9300_read_status got SDS mode 1f
[   71.770000] rtl9300_read_status: Port 27, SerDes is 9
[   71.770000] rtl9300_read_status got SDS mode 1f
[   72.800000] rtl9300_read_status: Port 27, SerDes is 9
[   72.800000] rtl9300_read_status got SDS mode 1f
[   73.850000] rtl9300_read_status: Port 27, SerDes is 9
[   73.850000] rtl9300_read_status got SDS mode 1f

```

u1f35c commented:
Not sure if user comments are wanted on this or not, but 2 from my initial testing:

The RTL8226 PHY code is forked here, rather than using the existing in-kernel driver. It's then lacking at least the support for reporting the 2500Mb/s modes via ethtool:

```
root@OpenWrt:/# ethtool lan10
Settings for lan10:
	Supported ports: [ TP	 MII ]
	Supported link modes:   10baseT/Half 10baseT/Full
	                        100baseT/Half 100baseT/Full
	                        1000baseT/Half 1000baseT/Full
	Supported pause frame use: Symmetric Receive-only
	Supports auto-negotiation: Yes
	Supported FEC modes: Not reported
	Advertised link modes:  10baseT/Half 10baseT/Full
	                        100baseT/Half 100baseT/Full
	                        1000baseT/Half 1000baseT/Full
	Advertised pause frame use: Symmetric Receive-only
	Advertised auto-negotiation: Yes
	Advertised FEC modes: Not reported
	Speed: 2500Mb/s
	Duplex: Full
	Auto-negotiation: on
	Port: Twisted Pair
	PHYAD: 25
	Transceiver: external
	MDI-X: Unknown
	Supports Wake-on: d
	Wake-on: d
	Link detected: yes
```

The link does seem to come up at 2500Mb/s and the remote end reports that speed is advertised, so this is more a diagnostic concern.

Secondly, adding an additional IP address to the switch interface doesn't seem to work:

```
root@OpenWrt:/# nft flush ruleset
root@OpenWrt:/# ip addr add 192.168.100.200/24 dev switch.1
root@OpenWrt:/# ip route
192.168.1.0/24 dev switch.1 proto kernel scope link src 192.168.1.1 offload 
192.168.100.0/24 dev switch.1 proto kernel scope link src 192.168.100.200 
root@OpenWrt:/# ping 192.168.100.100
PING 192.168.100.100 (192.168.100.100): 56 data bytes
^C
--- 192.168.100.100 ping statistics ---
7 packets transmitted, 0 packets received, 100% packet loss
root@OpenWrt:/# arp
IP address       HW type     Flags       HW address            Mask     Device
192.168.100.100  0x1         0x2         80:3f:5d:05:34:9c     *        switch.1
```

Not sure if there's something extra I need to be doing here, but I can see the packets getting out ok, ARP obviously works, but all IP seems to just get dropped.

oliv3r commented:
> Not sure if user comments are wanted on this or not, but 2 from my initial testing:

They are most defiantly appreciated :) I'm right now fighting with the RTL8226, which was working in birger's patches initially, so I did something wrong here, gotta figure out what now first.

> 
> The RTL8226 PHY code is forked here, rather than using the existing in-kernel driver. It's then lacking at least the support for reporting the 2500Mb/s modes via ethtool:

Was there an 8226 driver in the v5.10 kernel? Anyway, yes. These patches make a bit of a mess with regard to that. But splitting and cleaning all of that will take many months. So for now, it just follows the 'realtek' approach if you will. Not ideal, but better then nothing.

> 
> ```
> root@OpenWrt:/# ethtool lan10
> Settings for lan10:
> 	Supported ports: [ TP	 MII ]
> 	Supported link modes:   10baseT/Half 10baseT/Full
> 	                        100baseT/Half 100baseT/Full
> 	                        1000baseT/Half 1000baseT/Full
> 	Supported pause frame use: Symmetric Receive-only
> 	Supports auto-negotiation: Yes
> 	Supported FEC modes: Not reported
> 	Advertised link modes:  10baseT/Half 10baseT/Full
> 	                        100baseT/Half 100baseT/Full
> 	                        1000baseT/Half 1000baseT/Full
> 	Advertised pause frame use: Symmetric Receive-only
> 	Advertised auto-negotiation: Yes
> 	Advertised FEC modes: Not reported
> 	Speed: 2500Mb/s
> 	Duplex: Full
> 	Auto-negotiation: on
> 	Port: Twisted Pair
> 	PHYAD: 25
> 	Transceiver: external
> 	MDI-X: Unknown
> 	Supports Wake-on: d
> 	Wake-on: d
> 	Link detected: yes
> ```
> 
> The link does seem to come up at 2500Mb/s and the remote end reports that speed is advertised, so this is more a diagnostic concern.

I'm not much surprised. If I can spot the ehttool issue, I'll fix it; but I won't deep-dive into it however. Should be simple generally though.

> 
> Secondly, adding an additional IP address to the switch interface doesn't seem to work:
> 
> ```
> root@OpenWrt:/# nft flush ruleset
> root@OpenWrt:/# ip addr add 192.168.100.200/24 dev switch.1
> root@OpenWrt:/# ip route
> 192.168.1.0/24 dev switch.1 proto kernel scope link src 192.168.1.1 offload 
> 192.168.100.0/24 dev switch.1 proto kernel scope link src 192.168.100.200 
> root@OpenWrt:/# ping 192.168.100.100
> PING 192.168.100.100 (192.168.100.100): 56 data bytes
> ^C
> --- 192.168.100.100 ping statistics ---
> 7 packets transmitted, 0 packets received, 100% packet loss
> root@OpenWrt:/# arp
> IP address       HW type     Flags       HW address            Mask     Device
> 192.168.100.100  0x1         0x2         80:3f:5d:05:34:9c     *        switch.1
> ```
> 
> Not sure if there's something extra I need to be doing here, but I can see the packets getting out ok, ARP obviously works, but all IP seems to just get dropped.

Is that using initramfs? Others have reported the same. Flashing the build 'before my hackery' only works when booting from flash. More annoyingly, You can't even change the proto/ip address after it was set during boot. I do a quick `uci set network.lan.proto=dhcp` during dropbear key generation; and then dhcp works. Weirdly, I do see dhcp requests. There was some discussion on the forum about this however.

u1f35c commented:
> > Not sure if user comments are wanted on this or not, but 2 from my initial testing:
> 
> They are most defiantly appreciated :)

Great. I'd love to see this platform (and the RTL93xx chips generally) supported properly, as I'm in the market for some multigig upgrades.

> I'm right now fighting with the RTL8226, which was working in birger's patches initially, so I did something wrong here, gotta figure out what now first.
> 
> > The RTL8226 PHY code is forked here, rather than using the existing in-kernel driver. It's then lacking at least the support for reporting the 2500Mb/s modes via ethtool:
> 
> Was there an 8226 driver in the v5.10 kernel? Anyway, yes. These patches make a bit of a mess with regard to that. But splitting and cleaning all of that will take many months. So for now, it just follows the 'realtek' approach if you will. Not ideal, but better then nothing.

Yes, `drivers/net/phy/realtek.c` is in 5.10.136 and has support for the RTL8226. One problem is that .features for the 8226 is defined as `PHY_GBIT_FEATURES` but when I tried to change to `get_features` to read the supported bit I hit problems trying to read from the PHY. I suspect I was either doing so too early or the `phy_read_paged` function wasn't the right way to go - I got `EIO` errors. For the advertising piece `rtl8226_read_status` needs enhanced to look at the LPADV bits, but as I didn't get over the initial features problem I didn't get to testing that bit yet.
 
> > Secondly, adding an additional IP address to the switch interface doesn't seem to work:
…
> > Not sure if there's something extra I need to be doing here, but I can see the packets getting out ok, ARP obviously works, but all IP seems to just get dropped.
> 
> Is that using initramfs? Others have reported the same. Flashing the build 'before my hackery' only works when booting from flash. More annoyingly, You can't even change the proto/ip address after it was set during boot. I do a quick `uci set network.lan.proto=dhcp` during dropbear key generation; and then dhcp works. Weirdly, I do see dhcp requests. There was some discussion on the forum about this however.

Yes, it's using the initramfs. I'll have to look at actually flashing something and see if that's better.

lucize commented:
had another run to see about the sfp+ and I have this

when I first connect it

```
[   98.810000] rtl9300_sds_tx_config SerDes 9, pre-amp enable 1, pre-amp val 2, main-amp 9, post-amp enable 1, post-amp val 2, impedance 8
[   98.940000] ------------[ cut here ]------------
[   98.950000] WARNING: CPU: 0 PID: 302 at drivers/net/phy/phy.c:1159 phy_start+0xc0/0xe0
[   98.950000] called from state NOLINK
[   98.960000] Modules linked in: nft_fib_inet nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack lzo nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_ipv6 nf_log_ipv4 nf_log_common nf_defrag_ipv6 nf_defrag_ipv4 lzo_rle lzo_decompress lzo_compress libcrc32c sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact crypto_acompress gpio_button_hotplug crc32c_generic
[   99.030000] CPU: 0 PID: 302 Comm: kworker/0:2 Not tainted 5.10.139 #0
[   99.040000] Workqueue: events_power_efficient sfp_timeout
[   99.040000] Stack : 806e0000 825a4c00 80363b2c 00000009 00000000 8006374c 806e0000 00000017
[   99.050000]         825a4c3c 8274dd4c 8274dd1c 800fbdc8 80363b2c 00000001 8274dcf0 20ea949c
[   99.060000]         00000000 00000000 8063db44 8274db70 ffffefff 00000000 00000000 ffffffea
[   99.070000]         00000244 8274db7c 00000244 806edb68 00000000 00000009 8274ddd0 80363b2c
[   99.080000]         00000009 00000000 80760000 fffffffe 00000018 8030f270 00000000 80f80000
[   99.090000]         ...
[   99.090000] Call Trace:
[   99.090000] [<8000699c>] show_stack+0x30/0x100
[   99.100000] [<800285d0>] __warn+0xc0/0x12c
[   99.100000] [<800286c8>] warn_slowpath_fmt+0x8c/0xac
[   99.110000] [<80363b2c>] phy_start+0xc0/0xe0
[   99.110000] [<80362524>] phylink_sfp_module_start+0x40/0x50
[   99.120000] [<8036fdec>] sfp_sm_main+0x420/0x588
[   99.130000] [<803709b0>] sfp_timeout+0x40/0x5c
[   99.130000] [<8003fbcc>] process_one_work+0x1f0/0x460
[   99.140000] [<8003ffbc>] worker_thread+0x180/0x5c4
[   99.140000] [<8004630c>] kthread+0x13c/0x144
[   99.150000] [<80001ab8>] ret_from_kernel_thread+0x14/0x1c
[   99.150000]
[   99.150000] ---[ end trace 2f5988e4d3ba0258 ]---

```

when I reconnect it

```
[  127.980000] sfp sfp-p12: module removed
[  130.850000] sfp sfp-p12: module CISCO-FINISAR    FTLX8571D3BCL-C2 rev A    sn FNSxxxxxWVN      dc 140705
[...]
[  138.840000] ------------[ cut here ]------------
[  138.850000] WARNING: CPU: 0 PID: 2455 at drivers/net/phy/phy.c:1115 phy_stop+0x138/0x164
[  138.860000] called from state HALTED
[  138.860000] Modules linked in: nft_fib_inet nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack lzo nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_ipv6 nf_log_ipv4 nf_log_common nf_defrag_ipv6 nf_defrag_ipv4 lzo_rle lzo_decompress lzo_compress libcrc32c sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact crypto_acompress gpio_button_hotplug crc32c_generic
[  138.930000] CPU: 0 PID: 2455 Comm: netifd Tainted: G        W         5.10.139 #0
[  138.940000] Stack : 80770000 80665238 00000000 00000000 828f3ca4 80f80000 806e0000 8063db44
[  138.950000]         827d6c20 806e7e27 80f832a4 00000997 7fca0d60 00000001 828f3c58 0b0318f9
[  138.960000]         00000000 00000000 8063db44 828f3af8 ffffefff 00000000 00000000 ffffffea
[  138.970000]         000002c7 828f3b04 000002c7 806edb68 80000000 00000009 828f3d38 803655c4
[  138.980000]         00000009 7fca0d60 77e4b000 77e4b000 00000018 8030f270 00000000 80f80000
[  138.990000]         ...
[  138.990000] Call Trace:
[  138.990000] [<8000699c>] show_stack+0x30/0x100
[  139.000000] [<800285d0>] __warn+0xc0/0x12c
[  139.000000] [<800286c8>] warn_slowpath_fmt+0x8c/0xac
[  139.010000] [<803655c4>] phy_stop+0x138/0x164
[  139.010000] [<8035f814>] phylink_stop+0x44/0xd0
[  139.020000] [<80590338>] dsa_port_disable_rt+0x30/0xa8
[  139.020000] [<80591ba4>] dsa_slave_close+0x24/0xf4
[  139.030000] [<803f0bdc>] __dev_close_many+0x94/0x104
[  139.030000] [<803fce58>] __dev_change_flags+0xd4/0x21c
[  139.040000] [<803fcfc8>] dev_change_flags+0x28/0x70
[  139.040000] [<80428c8c>] dev_ioctl+0x244/0x5c0
[  139.050000] [<803cde64>] sock_ioctl+0x3f0/0x57c
[  139.050000] [<8017a808>] sys_ioctl+0xe0/0xf8
[  139.060000] [<8000e68c>] syscall_common+0x34/0x58
[  139.060000]
[  139.070000] ---[ end trace 2f5988e4d3ba0259 ]---

```

oliv3r commented:
i just pushed a new version with things back to a working state for the 2.5g ethernet; the SFP one is something I'll need too see if I can reproduce it; but i only hvae a DAC cable ...

lucize commented:
still no joy for me.. :(
I use and usb3 to 2.5gb realtek adapter
also the annoying repeating message,
 

```
[   33.472841] rtl930x_switch_irq RX symbol errors: 000001c4
[   33.503330] switch: port 3(lan3) entered blocking state
[   33.509159] switch: port 3(lan3) entered forwarding state
[   33.533301] rtl93xx_phylink_mac_config port 2, mode 0, phy-mode: xgmii, speed 1000, link 1
[   33.542496] rtl93xx_phylink_mac_config SDS is -1
[   34.401089] rtl9300_read_status: Port 27, SerDes is 9
[   34.406766] rtl9300_read_status got SDS mode 1f
[   34.520849] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 1000, link 1
[   34.530403] rtl93xx_phylink_mac_config SDS is 6
[   34.535514] rtl83xx-switch switch@1b000000 lan9: Link is Up - 1Gbps/Full - flow control rx/tx
[   34.546321] rtl930x_switch_irq link faults: 09ffff00
[   34.551851] rtl930x_switch_irq RX symbol errors: 000001c4
[   34.577955] switch: port 9(lan9) entered blocking state
[   34.583847] switch: port 9(lan9) entered forwarding state
[   34.602453] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 1000, link 1
[   34.612003] rtl93xx_phylink_mac_config SDS is 6
[   34.803381] rtl83xx_fib_event: FIB_RULE ADD/DELL for IPv6 not supported
[   34.810817] rtl83xx_fib_event: FIB_RULE ADD/DELL for IPv6 not supported
[   35.433254] rtl9300_read_status: Port 27, SerDes is 9
[   35.438877] rtl9300_read_status got SDS mode 1f
[   35.536023] rtl93xx_phylink_mac_config port 26, mode 0, phy-mode: usxgmii, speed 10000, link 1
[   35.545647] rtl93xx_phylink_mac_config SDS is 8
[   35.550703] rtl83xx-switch switch@1b000000 lan11: Link is Up - 10Gbps/Full - flow control rx/tx

```

```
[  108.233244] rtl9300_read_status: Port 27, SerDes is 9
[  108.238870] rtl9300_read_status got SDS mode 1f
[  109.280452] rtl9300_read_status: Port 27, SerDes is 9
[  109.286134] rtl9300_read_status got SDS mode 1f
[  110.313254] rtl9300_read_status: Port 27, SerDes is 9
[  110.318878] rtl9300_read_status got SDS mode 1f
[  111.360571] rtl9300_read_status: Port 27, SerDes is 9
[  111.366242] rtl9300_read_status got SDS mode 1f

```

piaddic7ed commented:
> I use and usb3 to 2.5gb realtek adapter
> also the annoying repeating message,
Same issue here with RTL8156B to USB-C running Manjaro (Linux 6.0rc4)

oliv3r committed:
realtek: Remove specific targets from generic image/Makefile

There seems to be no reason to have the HP 1920 and Netgear switches
as part of the main Makefile. Move them to their specific submakefiles.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented:
@lucize I don't have 2.5G hardware :( So I can only do loopback, or 1G connections. The 1G connection (on the 2.5g port works great) Shortening them, also. I need some help from setup a 'second' bridge, so that I can move lan10 and lan11 into that second bridge, and do 10G -> 10G -> 2.5G -> 2.5 -> 1G which yes, is nasty, but it also means all my traffic would be cleanly jumping from port to port :)

piaddic7ed commented:
> @lucize I don't have 2.5G hardware :( So I can only do loopback, or 1G connections. 

I'm able to reproduce the issue with loop back connections, too. Unplug and re-plug the cable is enough to trigger the issue.

lucize commented:
looking at the other patch, it seems that the annoying message is just removed

```
	mode = rtl9300_sds_mode_get(sds_num);
-	pr_info("%s got SDS mode %02x\n", __func__, mode);
+	if (mode == 0x1f)  // SerDes "off", look at forced mode instead 

```

piaddic7ed commented:
@oliv3r your last commit seems to fix the 2.5G ports. Vlan issues are still present.

lucize commented:
maybe do I need t write it to flash? I'm using only the initramfs images

piaddic7ed commented:
> maybe do I need t write it to flash? I'm using only the initramfs images

I'm only using initramfs as well.

oliv3r commented:
I mostly use the initramfs only as well ...

On 14 September 2022 21:24:36 CEST, piaddic7ed ***@***.***> wrote:
>> maybe do I need t write it to flash? I'm using only the initramfs images
>
>I'm only using initramfs as well.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/openwrt/openwrt/pull/10453#issuecomment-1247206035
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

lucize commented:
I tried a pcie nic, also realtek, the same issue
I also tried a 10G rj45 sfp+ module (hot plug) and I get a loop off:
https://pastebin.com/ygi8y4gm

oliv3r committed:
base-files: Do not break on non-eth ports

When using OpenWRT with DSA and 'lan' ports, we could get an empty
`next_eth`. This is of course not desirable, as this causes `sh: out of
range` errors when trying to determine which one would be greater.

Fixes: b688bf83f9d6 ("base-files: rename ethernet devs on known boards")
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
base-files: Do not parse the output of ls

It is considered bad practice, to parse the output of ls, so lets not.

While here, transform the cryptic hard-to-read shell script, to more
easily readable shell script and adhere to shell best-practices in
general and passing shellcheck.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Use OpenWRT default switch naming

With the bug of the default name not working resolved in XXX, we should
also just use the preferred default OpenWRT naming scheme.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Do not parse `ls` output

It is considered very bad practise to parse the output of `ls` in
shell scripts. We can do it easily without, however we do get an
unsorted list. Due to the lack of 'human sort' in the default OpenWRT
busybox sort function, we have to hack a little bit around it.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Be more versatile in getting the MAC address

Most of the realtek targets get the MAC address/range pretty standard
places. The U-Boot environment or the U-Boot 'sys' environment. Instead
of trying to read directly from flash, get these things from variables
if possible. Only if this fails, we fall back to alternative means.

This also means that this is generic enough, to support most/all of our
switches, except the HP ones, which embed the MAC address weirdly in the
header.

This means that, we first try to get the MAC address from the
devicetree, potentially set by U-Boot.

Secondly, we check if we have the U-Boot env tools to ask both the
environments, which gives us a nice and clean way to get this data.

If this fails, we'll resolve to more aggressive methods by dumping the
data from the various flashes. We are more flexible in naming of the
second partition, which could be named either `env2` or `sysenv`. This
more aggressive method is also useful for initramfs mode, which usually
does not expose the environment. A little bit of a hack ;).

Finally, if we fail to get a valid MAC address, resort to a random MAC
address supplied by OpenWRT.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Be a bit more careful when parsing the MAC addresses

We should check our pre-requisites better when setting the interface MAC
addresses.

For one, when we don't have an end mac address for the LAN ports,
we should set the Local Addressing bit for the LAN ports.

Secondly, if the label MAC and lan MAC are identical, increment by one
to ensure we don't use the same MAC address.

One Edge-case (which is handled) is that if the range is shorter then
the number of ports, when we encounter this, any following mac address
will use Local Addressing.

Further more, lets follow the original line of reasoning on MAC
assignments. The MAC address on the label, supplied via `ethaddr` should
be assigned to eth0 and the switch bridge, and 'mac_start - mac_end', if
available used for the individual lan ports. If those are not supplied,
use Link Local Addressing.

While here, make sure the new implementation is readable, shellcheck
compliant shell scripting, rather then cryptic, optimized for shortest
size.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
base-files: Add MAC address validation helper

Rather then checking if a (mac address containing) variable is empty or
not, lets instead allow a user to validate a MAC address.

This function can be used in one of two ways. Either just called as a
function, where the return value (of grep) actually tells you if the MAC
address matches (0) or not (1). Alternatively, capture the output, the
MAC address matches (the MAC address is repeated) or not (empty).

Currently, this check only supports byte seperated mac addresses, but
could be extended to support word-dot seperated addresses as well. Since
this is however a format not really used by OpenWRT, lets keep it simple
for now.

The used regex, `^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]){2}$` can be put
into regexr.com or similar to get a nice textual explanation for
those curious.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
base-files: Use sed instead of grep and awk

We can use a regex (with sed) to obtain the mtd index instead of grep
and awk, which a sub-shell spawned less and a pipe less.

However, performance is not really the reason for this change. This
change, especially the global regex definition, allows us to also obtain
the partition size, erase block size to allow us to generate the
environment partition more generically.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
base-files: Add U-Boot environment generator helpers

With our previously introduced generic regular expression, we can also
use this to obtain the partition size and erase-block size. This can even
be leveraged to get the U-Boot environment parameters in one go.

The new main function can be used as follows

```
find_mtd_env /dev/mtd1|u-boot-env [0x0ffset]
```
This function will then result a string, that can be passed straight to
`ubootenv_add_uci_config` or `ubootenv_add_uci_sys_config`, though it is
probably recommended to check the result before passing it along,
allowing for graceful failure/skipping.

Not included, is the partition offset within the mtd device, because
sadly `/proc/mtd` does not expose this. If we wanted to, we could get it
from sysfs however in the future.

```sh
find_mtd_offset() {
	printf '0x%x' "$("/sys/class/mtd/mtd$(find_mtd_index "${1:?}")/offset")"
}
```

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Drop partition data duplication

We are currently duplicating the U-Boot environment partition
information for no apparent reason. We already have a single-source of
truth for this data, `/proc/mtd` which is runtime generated by the
kernel, thus the most accurate. Lets not duplicate devicetree's and
potentially introduce copy/paste bugs.

The new code gets the data at runtime from `/proc/mtd` but still keeps
the switch-case for boards that can not work with the runtime method.
The runtime method will only generate the appropriate entries for actually
existing partitions.

Not only does this simplyfy the code, it also makes adding devices
easier and lower effort.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

plappermaul commented:
@oliv3r @svanheule May I ask how you want to proceed with this PR? Comparing this with the gitlab spin off repo it looks like we won't be able to include a single bit of Birgers patches.

oliv3r committed:
realtek: Expose sfp parameters

We need to be able to calibrate SerDes, and link parameters for this.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Use SDK tx eye calibration tables

The SDK offers various calibration tables that they use as eye
calibration tables. These tables are not ideal, as they name the serdes
that they where intended for, with the function of the reference model,
never the less, it gives some meaning and structure to these magic
values.

These values should be set in the devicetree in the future, as they are
very much hardware dependent.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Calibrate TX SDS config for DAC cables

To configure the transmission serdes controller, we need to know first
if we are dealing with a DAC type, and secondly the length (longer DAC's
use the standard setting). This information is available on a DSA port's
PhyLink struct we can use, but have to pass a long.

This does not yet fix all DAC lengths, as the RX calibration is still
needed, but that is far from trivial.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented:
Once things are actually properly working; and getting some positive test results, I'll re-push birgers patches (the ones in my branch on gitlab, that have been a little bit cleaned, ammended and refactored, to this branch for reference _only_. The idea was, to then take small subsections of patches, and offer those as individual MR's. But I can't do that until I know that everything works (or at least that worked on master).

plappermaul commented:
Understood. What I'm interested in is a clean/easy build environment for a Zyxel 1010. Bootable plus basic-ethernet-working. So that I can do module development without building full images. And this without the need to find out and carry around patches that are not mainlined yet.

I already asked in chat but currently nobody seems to be online. So I repeat it here: What is the minimum requirment to get this working? Maybe you know a simple fix without waiting for this epic MR.

Borromini commented:
@oliv3r Just tried to apply this on top of a clean master branch, but I'm seeing the following. Could you rebase?

```
$ git am 10453.patch
Applying: kernel: backport gnu11 upgrade
.git/rebase-apply/patch:68: trailing whitespace.
 
.git/rebase-apply/patch:74: trailing whitespace.
 
.git/rebase-apply/patch:94: space before tab in indent.
        const char *p;
.git/rebase-apply/patch:95: space before tab in indent.
        while (*pat) {
.git/rebase-apply/patch:98: space before tab in indent.
                p = *pat++;
warning: squelched 35 whitespace errors
warning: 40 lines add whitespace errors.
Applying: realtek: Improve debugging information to aid development
error: patch failed: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/dsa.c:1689
error: target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/dsa.c: patch does not apply
error: patch failed: target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:1697
error: target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c: patch does not apply
Patch failed at 0002 realtek: Improve debugging information to aid development
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

oliv3r commented:
@Borromini sure, but I'm actually working on a different branch, https://gitlab.com/olliver/openwrt/openwrt/-/tree/dev/realtek-next so no guarantee's this branch is as functional as I intended.

@plappermaul I've started to offer the patches as individual fixes, only 1 of the 2 you need so far, did only the 'fixes' things first, as the other one is an 'improve' patch, though I think it also is a 'fix' as it does rtl930x enablement, anyway, I don't want to overwhelm people on this thread too much ;)

@svanheule I've started to split up patches in their own commits, so they are easier to review. Mind you, this is all still birger's work, and some of it is magic, so lets not get too nitpicky here, as I still see these net drivers as learning grounds for the real drivers later.

!10959
!10960
!10961
!10962
!10963

I'll do more patches later as those are new features really and cleanups (and we can do the cleanups in one MR ...

Neustradamus commented:
@oliv3r: Any progress on this PR?

Djfe commented:
before new effort makes sense: the realtek target requires some help right now.
Someone needs to help out and update the target to kernel 6.6.
If that doesn't happen then the target might be dropped when the release is branched of. Adding a new device only makes sense if the realtek target stays supported.
https://github.com/openwrt/openwrt/pull/16052
http://lists.openwrt.org/pipermail/openwrt-devel/2024-August/043087.html

oliv3r commented:
I'm fully aware :p but I'm now also curious about the new mainlining effort. If this DSA patch gets surfaced then I'm curious. Though I must admit I am a bit sad that they are not involved here with openwrt; and just started to push rtl9xxx stuff into mainline. Which makes me wonder if it even works, because afaik rtl93xx barely works without any patches, and it heavily relies on u-boot as well ...

Djfe commented:
interesting 🤔
by "they" you mean Realtek?

oliv3r commented:
> interesting 🤔 by "they" you mean Realtek?

https://github.com/openwrt/openwrt/pull/16052#issuecomment-2285565687 no, alliedtelesis apparently.

Neustradamus commented:
@oliv3r: What is needed to be ready for merging?

oliv3r commented:
> @oliv3r: What is needed to be ready for merging?

This one? I bet a ton of work; however I need to diff this dts with the other one offered just to see if the other one 'makes sense', if that makes sense :)
